File: partiql-plan/src/main/java/org/partiql/plan/rel/RelJoin.java
package org.partiql.plan.rel;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.JoinType;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;
import org.partiql.plan.rex.Rex;

import java.util.List;

/**
 * Logical theta-join abstract base class.
 */
public abstract class RelJoin extends RelBase {

    /**
     * Creates a new {@link RelJoin} instance.
     *
     * @param left left input (operand 0)
     * @param right right input (operand 1)
     * @param condition join condition
     * @param joinType join type
     * @return new {@link RelJoin} instance
     */
    @NotNull
    public static RelJoin create(@NotNull Rel left, @NotNull Rel right, @NotNull Rex condition, @NotNull JoinType joinType) {
        return new Impl(left, right, condition, joinType);
    }

    /**
     * Gets the left input.
     * @return left input (operand 0)
     */
    @NotNull
    public abstract Rel getLeft();

    /**
     * Gets the right input.
     * @return right input (operand 1)
     */
    @NotNull
    public abstract Rel getRight();

    /**
     * Gets the join type.
     * @return JoinType
     */
    @NotNull
    public abstract JoinType getJoinType();

    /**
     * Gets the join condition.
     * @return the join condition.
     */
    @NotNull
    public abstract Rex getCondition();

    @NotNull
    @Override
    protected RelType type() {
        throw new UnsupportedOperationException("compute join type");
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        Operand c0 = Operand.single(getLeft());
        Operand c1 = Operand.single(getRight());
        return List.of(c0, c1);
    }

    @Override
    public <R, C> R accept(@NotNull OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitJoin(this, ctx);
    }

    @NotNull
    public abstract RelJoin copy(@NotNull Rel left, @NotNull Rel right);

    @NotNull
    public abstract RelJoin copy(@NotNull Rel left, @NotNull Rel right, @NotNull Rex condition, @NotNull JoinType joinType);

    private static class Impl extends RelJoin {

        private final Rel left;
        private final Rel right;
        private final Rex condition;
        private final JoinType joinType;

        private Impl(Rel left, Rel right, Rex condition, JoinType joinType) {
            this.left = left;
            this.right = right;
            this.condition = condition;
            this.joinType = joinType;
        }

        @NotNull
        @Override
        public Rel getLeft() {
            return left;
        }

        @NotNull
        @Override
        public Rel getRight() {
            return right;
        }

        @NotNull
        @Override
        public Rex getCondition() {
            return condition;
        }

        @NotNull
        @Override
        public JoinType getJoinType() {
            return joinType;
        }

        @NotNull
        @Override
        public RelJoin copy(@NotNull Rel left, @NotNull Rel right) {
            return new Impl(left, right, condition, joinType);
        }

        @NotNull
        @Override
        public RelJoin copy(@NotNull Rel left, @NotNull Rel right, @NotNull Rex condition, @NotNull JoinType joinType) {
            return new Impl(left, right, condition, joinType);
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rel/RelJoin.kt
package org.partiql.plan.rel

import org.partiql.plan.JoinType
import org.partiql.plan.rex.Rex

/**
 * TODO DOCUMENTATION
 */
public interface RelJoin : Rel {

    public fun getLeft(): Rel

    // TODO REMOVE ME TEMPORARY – https://github.com/partiql/partiql-lang-kotlin/issues/1575
    public fun getLeftSchema(): RelType?

    public fun getRight(): Rel

    // TODO REMOVE ME TEMPORARY – https://github.com/partiql/partiql-lang-kotlin/issues/1575
    public fun getRightSchema(): RelType?

    public fun getCondition(): Rex?

    public fun getJoinType(): JoinType

    override fun getChildren(): Collection<Rel> = listOf(getLeft(), getRight())

    override fun isOrdered(): Boolean = false

    override fun <R, C> accept(visitor: RelVisitor<R, C>, ctx: C): R = visitor.visitJoin(this, ctx)
}

/**
 * Default [RelJoin] implementation.
 */
internal class RelJoinImpl(
    left: Rel,
    right: Rel,
    condition: Rex?,
    joinType: JoinType,
    leftSchema: RelType?,
    rightSchema: RelType?,
) : RelJoin {

    // DO NOT USE FINAL
    private var _left = left
    private var _right = right
    private var _condition = condition
    private var _joinType = joinType
    private var _leftSchema = leftSchema
    private var _rightSchema = rightSchema

    private var _children: List<Rel>? = null

    override fun getLeft(): Rel = _left

    override fun getRight(): Rel = _right

    override fun getCondition(): Rex? = _condition

    override fun getJoinType(): JoinType = _joinType

    override fun getLeftSchema(): RelType? = _leftSchema

    override fun getRightSchema(): RelType? = _rightSchema

    override fun getChildren(): Collection<Rel> {
        if (_children == null) {
            _children = listOf(_left, _right)
        }
        return _children!!
    }

    override fun getType(): RelType {
        TODO("Not yet implemented")
    }

    override fun isOrdered(): Boolean = false

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RelJoin) return false
        if (_left != other.getLeft()) return false
        if (_right != other.getRight()) return false
        if (_condition != other.getCondition()) return false
        if (_joinType != other.getJoinType()) return false
        return true
    }

    override fun hashCode(): Int {
        var result = 1
        result = 31 * result + _left.hashCode()
        result = 31 * result + _right.hashCode()
        result = 31 * result + _condition.hashCode()
        result = 31 * result + _joinType.hashCode()
        return result
    }
}
